### PR TITLE
Added support for regression prediction, ie raw values instead of cla…

### PR DESCRIPTION
**Pull request by charles-r-earp**
…ssification, as a minor modification of fine_tune_ae().

### CODE REVIEW

**Comment by charles-r-earp** [comment]
I just used the auto_encoder code and instead of using the samples as the target output, used an additional set of iterators. 

This is the test code. 

#include <dll/network.hpp>
#include <dll/neural/dense_layer.hpp>

#include <iostream>
#include <vector>
#include <random>

int main() {
  using network_t = dll::dbn_desc< 
    dll::network_layers<
      dll::dense_layer<2, 1>
    >
  >::network_t; 
  int nsamples = 100;
  std::default_random_engine generator;
  std::uniform_real_distribution<float> distribution(-10, 10);
  std::vector<etl::fast_dyn_matrix<float, 2>> in_data(nsamples);
  std::vector<etl::fast_dyn_matrix<float, 1>> out_data(nsamples);
  auto line = [](float x){
    return x + 1;
  };
  float x, y;
  for(int i=0; i<nsamples; ++i) {
    in_data[i][0] = x = distribution(generator);
    in_data[i][1] = y = distribution(generator);
    out_data[i][0] = line(x) > y ? 1 : 0;
  }
  std::unique_ptr<network_t> net = std::make_unique<network_t>();
  net->fine_tune_reg(in_data.cbegin(), in_data.cend(), out_data.cbegin(), out_data.cend(), 1);
  std::cout << "dbn_regression success" << std::endl;
  return 0;
}

**Comment by wichtounet** [comment]
Hi, 

Thank you very much for this Pull Request. Sorry for the delay in answering, I'm not very active during the holidays. 

Everything seems in order. 

If I understand correctly: This enable to train regression models by using the same kind of generator as the auto encoder, but not enforcing auto-encoder model. Right ?

Thanks

Baptiste

File: include/dll/dbn_impl.hpp [hunk 1]
```diff
@@ -1594,6 +1594,69 @@ struct dbn final {
...
+     * \param max_epochs The maximum number of epochs to train the network for.
+     * \return The final classification error
+     */
+    template <typename Generator, cpp_enable_iff(is_generator<Generator>)>
+    weight fine_tune_reg(Generator& generator, size_t max_epochs) {
+        dll::auto_timer timer("net:train:ft:reg");
+
+        validate_generator(generator);
+
+        //cpp_assert(dll::input_size(layer_get<0>()) == dll::output_size(layer_get<layers - 1>()), "The network is not build //as an autoencoder");
```
Comment (wichtounet) [inline include/dll/dbn_impl.hpp]: You can directly remove this instead of commenting

File: include/dll/dbn_impl.hpp [hunk 2]
```diff
@@ -1594,6 +1594,69 @@ struct dbn final {
...
+     * \param in_first Iterator to the first sample
+     * \param in_last Iterator to the last sample
+     * \param out_first Iterator to the first output
+     * \param out_last Iterator to the last output
+     * \param max_epochs The maximum number of epochs to train the network for.
+     * \return The final average loss
+     */
+    template <typename InIterator, typename OutIterator>
+    weight fine_tune_reg(InIterator&& in_first, InIterator&& in_last, OutIterator&& out_first, OutIterator&& out_last, size_t max_epochs) {
+        //return "";
```
Comment (wichtounet) [inline include/dll/dbn_impl.hpp]: Remove this as well

**Comment by charles-r-earp** [comment]
Correct. My only issue is the return value.

**Comment by wichtounet** [comment]
Seems fine now. 

What is you issue with the return value ? Seems correct to me.

**Comment by wichtounet** [comment]
If no more issues, I'll merge this now.

**Comment by wichtounet** [comment]
If you are interested, I added a few (dumb) tests for regressions: https://github.com/wichtounet/dll/blob/master/test/src/unit/reg.cpp

You can also see the configuration of the error using MSE rather than CCE.

**Pull request merged**